keps/sig-cluster-lifecycle/kubeadm/3929-no-cri-socket-annotation/README.md

```diff
@@ -0,0 +1,756 @@
+override this configuration during kubeadm join using the --config flag. This would
+eliminate the need for kubeadm to store CRI socket configuration on each node, and
+instead rely on the Kubernetes configuration files for specifying this information.
+
+## Motivation
+
+Currently, kubeadm adds a CRI socket annotation to the Node object during the
+"init upload-config" phase, which specifies the endpoint for the CRI that is being
+used by the kubelet on each node. This annotation is persistent on the Node object,
+even if the kubelet is updated or the CRI is changed.
+
+After migration of container runtime endpoint flag to kubelet config, we can set
+cri socket in kubelet configuration.
+
+### Goals
+
+- Remove the use of CRI socket annotation on Node object
+- We will prioritize respecting the local setting over the global one.
+- [Not decided yet] For further node customized kubelet configuration, it can be saved locally
+  on disk with file path `/var/lib/kubelet/kubeadm-config.yaml`. (If not used, I will move it to Non-Goals)
+
+### Non-Goals
+
+- Update the CRI socket annotation on Node object to be the latest
+
+## Proposal
+
+1. init: upload a global kubelet configuration with cri socket
+2. join: can override it using --config
+3. upgrade: re-download global one, but should use local kubelet configuration firstly
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### init: upload a global kubelet configuration with cri socket
+
+- `kubeadm init` will not add the annotation to node.
+- `kubeadm init` will check the customized `--config` at first and if no cri socket is set, it will
+  auto-detect it and save it global configuration and local as well.
+
+### join: can override it using --config
+
+- `kubeadm join` will not add the annotation to node.
+- `kubeadm join` will download the kubelet configuration from apiserver and the customized `--config`
+  at first and auto-detect will work only if not set. Auto-detect may log a warning message if it may
+  be misconfigured and log a general debug log if there is multi CRI-sockets.
+
+### upgrade: re-download global one, but should use local kubelet configuration firstly
+
+- `kubeadm upgrade` will download the kubelet configuration from apiserver and respect local one.
+
+### Proposal 1: respect a list of configuration in local kubelet configuration, and in v1.27, CRI socket is the only one
+
+During `kubeadm ugprade`, kubeadm will read the local kubelet configuration in `/var/lib/kubelet/config.yaml`.
+kubeadm also download the kubelet configuration from configmap and replace the `containerRuntimeEndpoint` and
+`imageServiceEndpoint`(This maybe empty and I prefer to respect it as well) with the local configuration.
+
+A node-specific kubelet configuration list should be maintained in kubeadm code.
```

Comment:
I mean that we can maintain it in `cmd/kubeadm/app/constants/constants.go` or somewhere.

```
nodeSpecificKubeletConfigurations =  map[string]string{
	"containerRuntimeEndpoint": "the endpoint of container runtime service",
	"imageServiceEndpoint": "the endpoint of container image service",
	"providerID": "sets the unique ID of the instance that an external provider",
}
```

The list can be extended and I am not sure if the list should be modified by user in a global config `nodeSpecificKubeletConfiugrations`, that means a new array in ClusterConfiguration. The default value can be `containerRuntimeEndpoint` and `imageServiceEndpoint`.

Reply:
i think we can take the patches approach - any locally stored config file can act as a patch over the config from kubelet-config cm. but we may only consider some fields from the kubeletconfiguration v1beta1. this becomes tricky when v1beta1 is replaced by e.g. v1

Reply:
> any locally stored config file can act as a patch over the config from kubelet-config cm

this can also be tricky in some circumstances where we don't want to patch a field e.g. during upgrade, but hopefully it's not an issue

Reply:
If we use the `--patches` feature here to add a default patch dir on node, this can be done in this way.

However, if users want to specify a new patch dir, he has to merge the default patch files or content to the new one. That would be a little weird.

Reply:
i meant a generic patches approach where a local config file overrides the global config downloaded from the cm. not the  --patches feature.

this local file must be stored somewhere. perhaps in the same dir as config.yaml, but called config-instance.yaml. IMO it has a number of tricky aspects that need to be covered in the design doc for init, join, upgrade.